metricbeat/internal/metrics/opt.go

```diff
@@ -0,0 +1,67 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package metrics
+
+// OptUint is a wrapper for "optional" types, with the bool value indicating
+// if the stored int is a legitimate value.
+type OptUint struct {
+	exists bool
+	value  uint64
+}
+
+// NewUint returns a new OptUint wrapper
+func NewUint() OptUint {
+	return OptUint{
+		exists: false,
+		value:  0,
+	}
+}
+
+// NewUintFrom returns a new OptUint wrapper with a given int
+func NewUintFrom(i uint64) OptUint {
+	return OptUint{
+		exists: true,
+		value:  i,
+	}
+}
+
+// None marks the Uint as not having a value.
+func (opt *OptUint) None() {
+	opt.exists = false
+}
+
+// Exists returns true if the underlying value is valid
+func (opt OptUint) Exists() bool {
+	return opt.exists
+}
+
+// Some Sets a valid value inside the OptUint
+func (opt *OptUint) Some(i uint64) {
```

Comment:
I find this terminology confusing. I am not used at all with Rust :) but `Some` there seems to be like the constructor, here it is some kind of setter.

If we follow Rust semantics, I would prefer to don't have setters, I would prefer to have only builders that both return an `OptUint`, that can be directly assigned or compared. And two methods to check if they are none or some, something such as:
* `func NoneOptUint() OptUint { return OptUint{valid:false}}`
* `func SomeOptUint(i uint64) OptUint { return OptUint{valid:true, value:i}}`.
* `func (o OptUint) IsNone() { return !o.valid }`
* `func (o OptUint) IsSome() { return o.valid }`
* Well, and something to get the value, such as `func (o OptUint) Value() (uint64, bool) { return o.value, o.valid }`

If we go the setters way, then I would use something like `IsValid()` instead of `None()`, and  `Set()` instead of `Some()`.

Reply:
Agreed, I hacked up the API fairly quickly, the naming is a little weird.

Reply:
Although the type is inspired of typical `Option` in `Haskell`, `Rust`, `Scala`, ... we maybe don't want to use their nomenclature.

Proposal:

The zero value is always a `None`. In general `var x OptInt` is equivalent to `x := NoneOptUint()`. That is, I would not even introduce a `NoneOptUint`. The zero value, is already a valid value. This leaves us with a constructor that "sets" the optional to an actual value. e.g. `OptUintWith(x uint) OptUint`.

In go stdlib a common name to check for uninitialized values is `IsZero()`. e.g. `(time.Time).IsZero()` or  `reflex.IsZero()`. Other names like `IsValid` or `IsNil` don't really fit here. A single predicate might do it, as `IsSome() == !IsNone()` always.

Regarding getters/status checks we might want:
- `Get() uint` (or `Read`). `Get` always returns the value part. If you want to check the value is not set use `IsZero()`.
- `GetOr(or uint) uint`. `GetOr` returns the internal value, if the optional is not zero, otherwise the value given by `or`.
- `Take() (uint, bool)`: Returns the internal value + status, plus sets the optional to its zero value.  A call to `IsZero()` always returns true after `Take()` has been invoked.

Given that we have `OptUintWith`, we don't really need a setter. We can already assign by value.